core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/detach/StackDetachService.java

```diff
@@ -0,0 +1,75 @@
+package com.sequenceiq.cloudbreak.core.flow2.stack.detach;
+
+import static com.sequenceiq.cloudbreak.util.Benchmark.measure;
+
+import java.util.Date;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.Cluster;
+import com.sequenceiq.cloudbreak.logger.MDCBuilder;
+import com.sequenceiq.cloudbreak.service.cluster.ClusterService;
+import com.sequenceiq.cloudbreak.service.stack.StackService;
+
+/**
+ * This class detaches the stack from the environment.
+ * <p>
+ * It is achieved by renaming name and stack CRN. CLuster associated with the stack is also renamed as well.
+ * Before this operation is performed, the data lake associated with this stack is also renamed so that DL name and stack same are the same.
+ * <p>
+ */
+@Service
+public class StackDetachService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(StackDetachService.class);
+
+    private static final String DELIMITER = "-";
+
+    private static final String DETACHED = "detached";
+
+    @Inject
+    private ClusterService clusterService;
+
+    @Inject
+    private StackService stackService;
+
+    public Stack detachStack(Stack stack) {
+        MDCBuilder.buildMdcContext(stack);
+        stack.setDetached(true);
+        stack.setName(getNameToDetach(stack.getName()));
+        stack.setResourceCrn(getCrnToDetach(stack.getResourceCrn()));
+        Stack savedStack = measure(() -> stackService.save(stack),
+                LOGGER, "Stack save took {} ms for stack {}", stack.getName());
+        // While detaching the stack, name of the cluster is renamed to avoid conflicts.
+        Optional<Cluster> cluster = clusterService.findOneByStackId(savedStack.getId());
+        cluster.map(c -> {
+            String clusterName = c.getName() + DELIMITER + new Date().getTime();
+            c.setName(clusterName);
+            return c;
+        }).map(clusterService::save);
+        return stack;
+    }
+
+    String getNameToDetach(String stackName) {
+        return stackName + DELIMITER + DETACHED;
+    }
+
+    String getNameToAttach(String stackName) {
+        return StringUtils.removeEnd(stackName, DELIMITER + DETACHED);
+    }
+
+    String getCrnToDetach(String resourceCrn) {
```

Comment:
I would be also careful with this solution, here are some thoughts of mine which should be considered:
- CRN is used across CDP and this solution would increase the length of it. Since we are using it in databases and also in URLs, I am not sure if every table and every solution is prepared for the increased length
- with this, searching for DL CRN would result to return with logs regarding current DL and also logs regarding detached DL, I am not sure if it is a pro or a con

I have seen Peter asked in another comment if `detached` field is needed or not. The question here is the same I think. We can have this modified CRN (and modified name) with delimiter or we have can only the detached flag and optimize queries and constraints on database level for that field. One of them possibly would make possible to distinguish detached and attached datalake in CB service and in datalake service.